### PR TITLE
And a pass to convert vector.* operations to AArch64 inline assembly

### PR DESCRIPTION
**Pull request by asaadaldien**
This pass allows directly emitting machine instruction to vector operations. Its applied in the codegen pipeline right after the operation is already tiled into the target HW-vector size.

There is a single pattern implemented in this pass that converts `vector.contract` with matmul semantics into 4-sdot instructions and explicitly transpose the R.H.S. 

Using this pattern `matmul_i8_i8_i32_128x128x128` is 3.4x faster than not lowering to `sdot`. This also illustrate the importance of packed matmul layout which will remove the following costs:

- Efficient cache locality and less loads (4x less we are currently doing 8 loads with tiled layout only 2 )
- Explicitly transpose R.H.S

### CODE REVIEW

File: iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        rewriter.getContext(),
+        {int32x4VType, int32x4VType, int32x4VType, int32x4VType});
+
+    /// TODO(ataei): We an have a formatter like the c++ inline asm to
+    /// ssa-values to string names which will make the inline-assembly
+    /// statements more redable e.g :
+    /// sdot ${dstVec_0}.4s, ${lhs}.16b,${rhs}.4b[0]
+    auto packedResult = rewriter.create<LLVM::InlineAsmOp>(
+        loc, returnType,
+        ArrayRef<Value>({lhs, rhs, dstVec[0], dstVec[1], dstVec[2], dstVec[3]}),
```
Comment (asaadaldien) [inline iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp:102]: @nicolasvasilache Related to our conversation yesterday but down one level, the absent of 2D native LLVM `vector<4x4xi8>` constrained our IR to insert `shape_casts` everywhere. I was wondering as future ISA are going to work on 2D tiles and even few years old ISA like the bellow `sdot` have the concept of lanes in a vector it make sense to represent that in LLVM!

Comment (nicolasvasilache) [reply iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp:102]: First, glad to see this used, very nice :) !

Second, yes, Intel is the first to introduce a proper 2-D type for their AMX tiles, I expect this to generalize over time.

File: iree/compiler/Conversion/VectorToLLVM/Passes.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+/// Populates `patterns` to convert vector.contract op to a sequence
+/// of AArch64 inline assembly operations.
+void populateVectorContractToAArch64InlineAsm(
+    OwningRewritePatternList &patterns, MLIRContext *context);
+
+}  // namespace iree_compiler
+}  // namespace mlir
+
+#endif
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/VectorToLLVM/Passes.h:38]: Fix missing newline at end of file (here and everywhere else)

File: iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        rhsShape[1] != 4)
+      return failure();
+
+    Value inLhs = contractionOp.lhs();
+    Value inRhs = contractionOp.rhs();
+
+    if (accType.getElementType() != rewriter.getIntegerType(32)) {
+      return failure();
+    }
+    // TODO(ataei) : Remove this when linalg -> vector.contract isn't
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp:56]: If I remember correctly. It is explicitly adding sign extension cause thats the definition of the operation. So could you explain a bit more whats happening here.

File: iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    if (rhsType.getElementType() != rewriter.getIntegerType(8)) {
+      if (auto parentOp = cast<SignExtendIOp>(inRhs.getDefiningOp())) {
+        inRhs = parentOp.value();
+      } else {
+        return failure();
+      }
+    }
+    auto loc = contractionOp.getLoc();
+
+    auto dstVec =
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp:74]: Hard to parse this statement. Its actually strange that the type is `SmallVector<Value>`. My first read of this was that it would be `SmallVector<vector::ExtractOp>`. So maybe make the type explicit.
Nit: Avoid the llvm::map_range completely for just iterating over a seq, and just write a loop.

File: iree/compiler/Conversion/VectorToLLVM/test/vector_contract_to_aarch64_asm.mlir [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+            affine_map<(d0, d1, d2) -> (d0, d2)>,
+            affine_map<(d0, d1, d2) -> (d2, d1)>,
+            affine_map<(d0, d1, d2) -> (d0, d1)>
+        ], iterator_types = ["parallel", "parallel", "reduction"], kind = #vector.kind<add>
+    } %3, %4, %2 : vector<4x4xi32>, vector<4x4xi32> into vector<4x4xi32>
+    vector.transfer_write %5, %dst[%c0, %c0] {masked = [false, false]}: vector<4x4xi32>, memref<4x4xi32>
+    return
+}
+// CHECK-LABEL: vector_matmul_to_aarch64_asm_vec_dot
+// CHECK: %[[RES_2D:.+]] = constant dense<0> : vector<4x4xi32>
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/VectorToLLVM/test/vector_contract_to_aarch64_asm.mlir:23]: Lets try some consistent FileCheck writing in IREE (at least for codegen parts). I'd prefer to 

1) Align all the `:`.
2) Use Check-DAG wherever a sequence of operations dont have a specific order.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Looks good! Just a few nits

File: iree/compiler/Conversion/VectorToLLVM/Passes.h (hunk 2 above)
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/VectorToLLVM/Passes.h:38]: Still missing newline.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp (hunk 3 above)
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp]: Still would like to understand whats happening here. This seems to be overriding the definition of the op. So want to understand this is fine.

Comment (asaadaldien) [reply iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp]: The op is defined to sign extend the i8 operands to  i32 so `vector.contract` operands are always i32:
https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Linalg/IR/LinalgNamedStructuredOpsSpec.tc#L30
here we folded the sexti and directly get the i8 operands which what the `sdot` instruction needs.

File: iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        rhsShape[1] != 4)
+      return failure();
+
+    Value inLhs = contractionOp.lhs();
+    Value inRhs = contractionOp.rhs();
+
+    if (accType.getElementType() != rewriter.getIntegerType(32)) {
+      return failure();
+    }
+    // TODO(ataei) : Currently linalg op lowering sign extends all operands to
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp:56]: This TODO is misleading. Just drop this. The operations semantics is defining the exact computation needed, and you are looking for that pattern. So it is consistent. The comment was misleading to me.

File: iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    // i32 we need to remove this when linalg -> vector.contract isn't
+    // explicitly adding sexti ops.
+    if (lhsType.getElementType() != rewriter.getIntegerType(8)) {
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp:59]: Nit: Make a single lambda that returns `Optional<Value>` and use that for getting the value here.

```
auto get8BitValue = [&](Value v) -> Optional<Value> {
  if (auto parentOp = v.getDefiningOp<SignExtendIOp>()) {
    return parentOp.value();
  }
  return llvm::None;
}
auto lhs8Bit = get8BitValue(inLhs);
auto rhs8Bit = get8BitValue(inRhs);
if (!lhs8Bit || !rhs8Bit) return failure();
```

In any case, you should not use `cast<SignExtendIOp>(...)`, use `.getDefiningOp<SignExtendIOp>()`.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Found a small bug.

File: iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp (hunk 6 above)
Comment (asaadaldien) [reply iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp:56]: My understanding this is done so lowering to loops works because scalar ops don't support mixed integer bits. On the other hand `vector.contract` does accept mixed integer math. I agree with you the comment is misleading and I will remove it, but want to anchor somehow the fact that `linalg.matmul_i8_i8_i32` shoudn't lower to `vector.contact` with i32 sign extended operands :)

File: iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp (hunk 7 above)
Comment (asaadaldien) [reply iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp]: Thanks! didn't now getDefiningOp does have casting semantics :)

**Review by stellaraccident — COMMENTED** [review]
Really glad to see this shaping up!

File: iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+    if (accType.getElementType() != I32Type) {
+      return failure();
+    }
+
+    auto getI8Value = [&](Value v) -> Value {
+      if (auto parentOp = v.getDefiningOp<SignExtendIOp>()) {
+        if (parentOp.value().getType().cast<VectorType>().getElementType() !=
+            I8Type) {
+          return NULL;
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp:64]: nit: Strange to see `NULL`. Better to use `nullptr`.

**Review by MaheshRavishankar — APPROVED** [review]
